ambry-named-mysql/src/main/java/com/github/ambry/named/GetTransactionStateTracker.java

```diff
@@ -0,0 +1,62 @@
+/**
+ * Copyright 2021 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.named;
+
+import com.github.ambry.rest.RestServiceErrorCode;
+import com.github.ambry.rest.RestServiceException;
+import java.util.PriorityQueue;
+
+
+/**
+ * An implementation of the {@link TransactionStateTracker}. It determine if a request should be retried on another
+ * datacenter based on an exception received from the last datacenter tried (processFailure). If processFailure returned
+ * true, the next call to getDatacenter should return the datacenter to retry on.
+ */
+public class GetTransactionStateTracker implements TransactionStateTracker{
+  private String nextDatacenter;
+  private final PriorityQueue<String> remoteDatacenters;
+
+  /**
+   * Constructor for {@link GetTransactionStateTracker}.
+   * @param remoteDatacenters list of the remote datacenter names.
+   * @param localDatacenter the name of the local datacenter.
+   */
+  GetTransactionStateTracker(PriorityQueue<String> remoteDatacenters, String localDatacenter) {
+    this.remoteDatacenters = remoteDatacenters;
+    this.nextDatacenter = localDatacenter;
+  }
+
+  /**
+   * // if its something like a not found, maybe handling of delete in tthe future
+   * @param throwable The {@link Throwable} thrown in execution of retrying.
+   * @return {@code true} if the {@link Throwable} is {@link RestServiceException} and Get request return not found.
+   */
+  @Override
+  public boolean processFailure(Throwable throwable) {
+    boolean isRetriable = throwable instanceof RestServiceException && RestServiceErrorCode.NotFound.equals(
+        ((RestServiceException) throwable).getErrorCode());
+    if (!remoteDatacenters.isEmpty() && isRetriable) {
+      nextDatacenter = remoteDatacenters.poll();
+    } else if (remoteDatacenters.isEmpty()) {
+      //Retried all datacenters
+      return false;
+    }
+    return isRetriable;
+  }
+
+  @Override
+  public String getNextDatacenter() {
```

Comment:
Updated.

Reply:
Actually I give a second thought, for the new retry logic, since we are not using the scheduler for retry, so the retry logic will be running in the same thread with the previous try run, so it will hit the same cache right? in that case, do we still need to set volatile for the counter or nextDatacenter?

Reply:
Yeah I think you're right that without the delegation due to delay, the call for processFailure and getNextDatacenter will happen in the same callback created by the RetryExecutor.

I do wonder about if the next request fails, since the next RetryExecutor callback will be run from a different TransactionExecutor thread pool. This will likely be okay in practice but i'm not completely sure if its guaranteed (need to do more research)

Reply:
Discussed offline, no matter the executeTransaction submit the different thread to get container from mysql or not, the logic inside GetTransactionStateTracker will still be in the same thread, so the counter value will be always from the same cache.